// tstl/base/container/SetElementVector.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __read = (this && this.__read) || function (o, n) {
    var m = typeof Symbol === "function" && o[Symbol.iterator];
    if (!m) return o;
    var i = m.call(o), r, ar = [], e;
    try {
        while ((n === void 0 || n-- > 0) && !(r = i.next()).done) ar.push(r.value);
    }
    catch (error) { e = { error: error }; }
    finally {
        try {
            if (r && !r.done && (m = i["return"])) m.call(i);
        }
        finally { if (e) throw e.error; }
    }
    return ar;
};
Object.defineProperty(exports, "__esModule", { value: true });
//================================================================ 
/** @module std.base */
//================================================================
var VectorContainer_1 = require("./VectorContainer");
var ArrayIteratorBase_1 = require("../iterator/ArrayIteratorBase");
/**
 * @hidden
 */
var SetElementVector = /** @class */ (function (_super) {
    __extends(SetElementVector, _super);
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    function SetElementVector(associative) {
        var _this = _super.call(this) || this;
        _this.data_ = [];
        _this.associative_ = associative;
        return _this;
    }
    SetElementVector.prototype.nth = function (index) {
        return new SetElementVector.Iterator(this, index);
    };
    /**
     * @internal
     */
    SetElementVector._Swap_associative = function (x, y) {
        var _a;
        _a = __read([y.associative_, x.associative_], 2), x.associative_ = _a[0], y.associative_ = _a[1];
    };
    /* ---------------------------------------------------------
        ACCESSORS
    --------------------------------------------------------- */
    SetElementVector.prototype.associative = function () {
        return this.associative_;
    };
    return SetElementVector;
}(VectorContainer_1.VectorContainer));
exports.SetElementVector = SetElementVector;
(function (SetElementVector) {
    var Iterator = /** @class */ (function (_super) {
        __extends(Iterator, _super);
        function Iterator() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @inheritDoc
         */
        Iterator.prototype.source = function () {
            return this._Get_array().associative();
        };
        /**
         * @inheritDoc
         */
        Iterator.prototype.reverse = function () {
            return new ReverseIterator(this);
        };
        return Iterator;
    }(ArrayIteratorBase_1.ArrayIteratorBase));
    SetElementVector.Iterator = Iterator;
    var ReverseIterator = /** @class */ (function (_super) {
        __extends(ReverseIterator, _super);
        function ReverseIterator() {
            return _super !== null && _super.apply(this, arguments) || this;
        }
        /**
         * @hidden
         */
        ReverseIterator.prototype._Create_neighbor = function (base) {
            return new ReverseIterator(base);
        };
        return ReverseIterator;
    }(ArrayIteratorBase_1.ArrayReverseIteratorBase));
    SetElementVector.ReverseIterator = ReverseIterator;
})(SetElementVector = exports.SetElementVector || (exports.SetElementVector = {}));
exports.SetElementVector = SetElementVector;
