// tstl/exception/RuntimeError.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
//================================================================ 
/** @module std */
//================================================================
var Exception_1 = require("./Exception");
/**
 * Runtime Error.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var RuntimeError = /** @class */ (function (_super) {
    __extends(RuntimeError, _super);
    /**
     * Initializer Constructor.
     *
     * @param message The error messgae.
     */
    function RuntimeError(message) {
        return _super.call(this, message) || this;
    }
    return RuntimeError;
}(Exception_1.Exception));
exports.RuntimeError = RuntimeError;
/**
 * Overflow Error.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var OverflowError = /** @class */ (function (_super) {
    __extends(OverflowError, _super);
    /**
     * Initializer Constructor.
     *
     * @param message The error messgae.
     */
    function OverflowError(message) {
        return _super.call(this, message) || this;
    }
    return OverflowError;
}(RuntimeError));
exports.OverflowError = OverflowError;
/**
 * Underflow Error.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var UnderflowError = /** @class */ (function (_super) {
    __extends(UnderflowError, _super);
    /**
     * Initializer Constructor.
     *
     * @param message The error messgae.
     */
    function UnderflowError(message) {
        return _super.call(this, message) || this;
    }
    return UnderflowError;
}(RuntimeError));
exports.UnderflowError = UnderflowError;
/**
 * Range Error.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var RangeError = /** @class */ (function (_super) {
    __extends(RangeError, _super);
    /**
     * Initializer Constructor.
     *
     * @param message The error messgae.
     */
    function RangeError(message) {
        return _super.call(this, message) || this;
    }
    return RangeError;
}(RuntimeError));
exports.RangeError = RangeError;
exports.runtime_error = RuntimeError;
exports.overflow_error = OverflowError;
exports.underflow_error = UnderflowError;
exports.range_error = RangeError;
