// tstl/utility/Pair.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var hash_1 = require("../functional/hash");
var comparators_1 = require("../functional/comparators");
/**
 * Pair of two elements.
 *
 * @author Jeongho Nam <http://samchon.org>
 */
var Pair = /** @class */ (function () {
    /* ---------------------------------------------------------
        CONSTRUCTORS
    --------------------------------------------------------- */
    /**
     * Initializer Constructor.
     *
     * @param first The first element.
     * @param second The second element.
     */
    function Pair(first, second) {
        this.first = first;
        this.second = second;
    }
    /* ---------------------------------------------------------
        COMPARISON
    --------------------------------------------------------- */
    /**
     * @inheritDoc
     */
    Pair.prototype.equals = function (pair) {
        return comparators_1.equal_to(this.first, pair.first) && comparators_1.equal_to(this.second, pair.second);
    };
    /**
     * @inheritDoc
     */
    Pair.prototype.less = function (pair) {
        if (comparators_1.equal_to(this.first, pair.first) === false)
            return comparators_1.less(this.first, pair.first);
        else
            return comparators_1.less(this.second, pair.second);
    };
    /**
     * @inheritDoc
     */
    Pair.prototype.hashCode = function () {
        return hash_1.hash(this.first, this.second);
    };
    return Pair;
}());
exports.Pair = Pair;
exports.pair = Pair;
/**
 * Create a {@link Pair}.
 *
 * @param first The 1st element.
 * @param second The 2nd element.
 *
 * @return A {@link Pair} object.
 */
function make_pair(first, second) {
    return new Pair(first, second);
}
exports.make_pair = make_pair;
